Mathlib/RingTheory/Polynomial/Eisenstein/Generalized.lean

```diff
@@ -0,0 +1,259 @@
+  set q : ℤ [X] := X ^ 2 + 1 with hq_eq
+  let K := ℤ ⧸ P
+  have hq_monic : q.Monic := leadingCoeff_X_pow_add_one (by norm_num)
+  have hq_deg : (X ^ 2 + 1 : K[X]).natDegree = 2 := by
+    simp only [← C_1, Polynomial.natDegree_X_pow_add_C]
+  have hq_irr : Irreducible (q.map (algebraMap ℤ K)) := by
+    have : map (algebraMap ℤ K) q = X ^ 2 + 1 := by simp [q]
+    rw [this]
+    have mod3 : ∀ u : ZMod 3, u ^ 2 + 1 ≠ 0 := by decide
+    rw [Polynomial.irreducible_iff_roots_eq_zero_of_degree_le_three]
+    · apply Multiset.eq_zero_of_forall_not_mem
+      intro a
+      simp only [mem_roots', ne_eq, IsRoot.def, eval_add, eval_pow, eval_X, eval_one, not_and]
+      intro _
+      let u := Int.quotientSpanNatEquivZMod 3 a
+      have : a ^ 2 + 1 = (Int.quotientSpanNatEquivZMod 3).symm (u ^ 2 + 1) := by simp [u]
+      rw [this]
+      rw [← ne_eq, RingEquiv.map_ne_zero_iff (Int.quotientSpanNatEquivZMod 3).symm]
+      apply mod3
+    · rw [hq_deg]
+    · rw [hq_deg]; norm_num
+  set f : ℤ[X] := X ^ 4 - 10 * X ^ 2 + 1 with hf_eq
+  have hdeg_f : f.natDegree = 4 := by
+    simp [hf_eq]
+    conv_rhs => rw [← natDegree_X_pow (R := ℤ) 4]
+    apply natDegree_eq_of_degree_eq
+    rw [sub_add] --  degree_X_pow (R := ℤ)]
+    apply degree_sub_eq_left_of_degree_lt
+    apply lt_of_le_of_lt (degree_sub_le _ _)
+    rw [← map_ofNat C, degree_mul, degree_C, degree_X_pow]
+    simp; norm_num; norm_num
+  have hlC_f : f.Monic := by
+    suffices coeff (1 : ℤ[X]) 4 = 0 by
+      simp [Monic, leadingCoeff, hdeg_f, f, this]
+    rw [← C_1, coeff_C, if_neg (by norm_num)]
+  have hfq : f = q ^ 2 - 12 * q + 12 := by ring
+  apply generalizedEisenstein hP (K := K) hq_irr hq_monic (p := 2)
+    (by rw [hdeg_f]; norm_num) hlC_f
+  · simp only [hfq, Polynomial.map_sub, Polynomial.map_add, Polynomial.map_pow,
+      Polynomial.map_mul, ← map_ofNat C, Polynomial.map_C]
+    have : (algebraMap ℤ K) (12) = 0 := by
+      rw [IsScalarTower.algebraMap_eq ℤ (ℤ ⧸ P) K, RingHom.comp_apply]
+      convert map_zero _
+      simp only [Ideal.Quotient.algebraMap_eq P, eq_zero_iff_mem, P, Ideal.mem_span_singleton]
+      · norm_num
+      · exact Submodule.Quotient.instZeroQuotient P
+      · exact AddMonoidHomClass.toZeroHomClass
+    rw [this]
+    simp
+  · suffices f %ₘ q = 12 by
+      rw [this]
+      rw [← map_ofNat C, Polynomial.map_C, ne_eq, C_eq_zero, eq_zero_iff_mem]
+      simp [P, Ideal.span_singleton_pow, Ideal.mem_span_singleton]
+    rw [hfq]
+    rw [← modByMonicHom_apply, LinearMap.map_add]
+    convert zero_add _
+    · rw [← LinearMap.mem_ker, mem_ker_modByMonic hq_monic]
+      rw [pow_two, ← sub_mul]
+      apply dvd_mul_left
+    · symm
+      simp only [modByMonicHom_apply, Polynomial.modByMonic_eq_self_iff hq_monic, f, P]
+      apply Polynomial.degree_lt_degree
+      suffices q.natDegree = 2 by
+        simp only [this, natDegree_ofNat, f, P, K, q]
+        norm_num
+      rw [hq_eq]
+      simp only [← C_1, Polynomial.natDegree_X_pow_add_C]
```

Comment:
You can use `compute_degree!` here (it needs `import Mathlib.Tactic.ComputeDegree`).

Reply:
Cool!